pulsar/transaction.go

```diff
@@ -17,7 +17,53 @@
 
 package pulsar
 
+import (
+	"context"
+)
+
+// TxnState The state of the transaction. Check the state of the transaction before executing some operation
+// with the transaction is necessary.
+type TxnState int32
+
+const (
+	_ TxnState = iota
+	// Open The transaction in Open state can be used to send/ack messages.
+	Open
+	// Committing The state of the transaction will be Committing after the commit method is called.
+	// The transaction in Committing state can be committed again.
+	Committing
+	// Aborting The state of the transaction will be Aborting after the abort method is called.
+	// The transaction in Aborting state can be aborted again.
+	Aborting
+	// Committed The state of the transaction will be Committed after the commit method is executed success.
+	// This means that all the operations with the transaction are success.
+	Committed
+	// Aborted The state of the transaction will be Aborted after the abort method is executed success.
+	// This means that all the operations with the transaction are aborted.
+	Aborted
+	// Errored The state of the transaction will be Errored after the operation of transaction get a non-retryable error.
+	Errored
```

Comment:
I see the error code is `ERROR` in the Java client. Should we name it with `Error` here? `Errored` looks confusing. We use the `ed` and `ing` suffixes to differ whether a step is complete or ongoing.

Reply:
The Error is a conflict with `Error` in error.go.
```java
type Error struct {
	msg    string
	result Result
}
```

Reply:
We get an unretrievable error and then change the state of the transaction to `Errored`.
So the `Errored` is OK.

Reply:
> The Error is a conflict with Error in error.go.

Okay. Then I think it's better to add a prefix to these enum constants. e.g. `TxnState_Open`, `TxnState_Error`. Otherwise, it could be very easy to cause the naming conflicts. You can also see the generated code in `PulsarApi.pb.go`, the enum constants all have their prefixes like `CompressionType_` and `ProducerAccessMode_`.

Reply:
I just found https://github.com/apache/pulsar-client-go/pull/471 renames the error code so that the prefix `Result` has been removed for all `Result` constants. It's a bad breaking change and I don't know why the reviewers still approved.

But we should avoid the wrong design like that.

Reply:
Okay, I add a prefix to the state. Because the `TxnState_Open` does not conform to Go's specification, I change it to `TxnOpen`